### PR TITLE
Make outputs of NuGet nomination immutable

### PR DESCRIPTION
**Pull request by davkean**
In preparation for moving NuGet nomination to operation progress, I'm moving it to a data source/snapshot model. In preparation for that, made all outputs immutable.

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/Snapshots/ImmutablePropertyCollection.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.NuGet
+{
+    /// <summary>
+    ///     Abstract immutable collection that supports lookup by index and name.
+    /// </summary>
+    internal abstract class ImmutablePropertyCollection<T> : IEnumerable where T : class
```
Comment (davidwengier) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/Snapshots/ImmutablePropertyCollection.cs]: Any reason not to implement `IEnumerable<T>` instead?

Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/Snapshots/ImmutablePropertyCollection.cs]: Extra implementation that will never be called; the `IVsXXX` collections only use IEnumerable.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/ProjectRestoreInfoBuilder.cs [hunk 2]
```diff
@@ -119,21 +107,20 @@ private static IVsReferenceItems GetReferences(IImmutableDictionary<string, IImm
 
         private static IVsReferenceItems GetProjectReferences(IImmutableDictionary<string, IImmutableDictionary<string, string>> projectReferenceItems)
         {
-            IVsReferenceItems referenceItems = GetReferences(projectReferenceItems);
+            // NuGet expects each "<ProjectReference />" to have metadata item "ProjectFileFullPath" that represents
+            // the full path of the project reference, just use the built-in "FullPath" metadata for that.
+            IImmutableDictionary<string, IImmutableDictionary<string, string>> replacementProjectReferenceItems = projectReferenceItems;
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/ProjectRestoreInfoBuilder.cs]: Using a builder here would likely involve fewer allocations. Might not matter if there aren't likely to be many changes.

Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/ProjectRestoreInfoBuilder.cs]: This code is deleted in a latter branch anyway.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/ProjectRestoreInfoBuilder.cs [hunk 3]
```diff
@@ -26,8 +26,8 @@ internal static IVsProjectRestoreInfo Build(IEnumerable<IProjectVersionedValue<I
 
             string msbuildProjectExtensionsPath = null;
             string originalTargetFrameworks = null;
-            var targetFrameworks = new TargetFrameworks();
-            var toolReferences = new ReferenceItems();
+            var targetFrameworks = ImmutableDictionary.Create<string, IVsTargetFrameworkInfo>();
+            var toolReferences = ImmutableDictionary.Create<string, IVsReferenceItem>();
```
[the remark this answers is not in the document]
Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/ProjectRestoreInfoBuilder.cs]: Good call on comparer, I was just trying to maintain (incorrect) behavior, but I could have done that with an explicit comparer.

This class is basically thrown away in latest code I'm working in.

[the remark this answers is not in the document]
Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/ProjectRestoreInfoBuilder.cs]: I'm going to do this in a followup.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/ProjectRestoreInfoBuilder.cs (hunk 2 above)
Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/ProjectRestoreInfoBuilder.cs]: Going to do this in a followup.

**Pull request merged**